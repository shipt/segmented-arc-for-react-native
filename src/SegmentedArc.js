import React, { useState, useEffect, useRef, createContext } from 'react';
import { Animated, Easing, StyleSheet } from 'react-native';
import PropTypes from 'prop-types';
import { View } from 'react-native';
import Svg from 'react-native-svg';
import Segment from './components/Segment';
import Cap from './components/Cap';
import RangesDisplay from './components/RangesDisplay';

const SegmentedArcContext = createContext();

export const SegmentedArc = ({
  fillValue = 0,
  segments = [],
  filledArcWidth = 8,
  emptyArcWidth = 8,
  spaceBetweenSegments = 2,
  arcDegree = 180,
  radius = 100,
  animationDuration = 1000,
  isAnimated = true,
  animationDelay = 0,
  showArcRanges = false,
  middleContentContainerStyle = {},
  ranges = [],
  rangesTextColor = '#000000',
  rangesTextStyle = styles.rangeTextStyle,
  capInnerColor = '#28E037',
  capOuterColor = '#FFFFFF',
  alignRangesWithSegments = true,
  children
}) => {
  const [arcAnimatedValue] = useState(new Animated.Value(0));
  const animationRunning = useRef(false);

  if (segments.length === 0) {
    return null;
  }

  const totalArcs = segments.length;
  const totalSpaces = totalArcs - 1;
  const totalSpacing = totalSpaces * spaceBetweenSegments;

  const arcSegmentDegree = (arcDegree - totalSpacing) / totalArcs;
  const arcsStart = 90 - arcDegree / 2;

  const effectiveRadius = radius + Math.max(filledArcWidth, emptyArcWidth);
  const margin = 12;
  let svgHeight = effectiveRadius + margin * 1.5;
  const svgWidth = effectiveRadius * 2 + 2 * margin;
  const center = effectiveRadius + margin;

  if (arcDegree > 180) {
    const offsetAngle = (arcDegree - 180) / 2;
    const heightOffset = effectiveRadius * Math.sin(offsetAngle);
    svgHeight += heightOffset + 2 * margin + filledArcWidth;
  }

  const localMiddleContentContainerStyle = {
    position: 'absolute',
    alignSelf: 'center',
    top: filledArcWidth + margin + 16,
    width: radius + filledArcWidth * 2,
    height: radius,
    alignItems: 'center',
    overflow: 'hidden',
    ...middleContentContainerStyle
  };

  const _ensureDefaultSegmentScale = () => {
    const segmentsWithoutScale = segments.filter(segment => !segment.scale);
    const allocatedScale = segments.reduce((total, current) => total + (current.scale || 0), 0);
    const defaultArcScale = (1 - allocatedScale) / segmentsWithoutScale.length;
    segmentsWithoutScale.forEach(segment => (segment.scale = defaultArcScale));
  };

  const _ensureDefaultArcScale = () => {
    const segmentsWithoutArcDegreeScaleLength = segments.filter(
      segment => typeof segment.arcDegreeScale !== 'number'
    ).length;
    const totalProvidedArcDegreeScale = segments.reduce((acc, segment) => acc + (segment.arcDegreeScale ?? 0), 0);
    segments.forEach(segment => {
      if (typeof segment.arcDegreeScale !== 'number')
        segment.arcDegreeScale = (1 - totalProvidedArcDegreeScale) / segmentsWithoutArcDegreeScaleLength;
    });
  };

  let remainingValue = fillValue;

  _ensureDefaultSegmentScale();
  _ensureDefaultArcScale();

  let arcs = [];
  segments.forEach((segment, index) => {
    const arcDegreeScale = segment.arcDegreeScale;
    const previousSegmentEnd = !!index ? arcs[index - 1].end : arcsStart;
    const start = previousSegmentEnd + (!!index ? spaceBetweenSegments : 0);
    const end = (arcDegree - totalSpacing) * arcDegreeScale + start;

    const valueMax = 100 * segment.scale;
    const effectiveScaledValue = Math.min(remainingValue, valueMax);
    const scaledPercentage = effectiveScaledValue / valueMax;
    const filled = start + scaledPercentage * (end - start);
    remainingValue -= effectiveScaledValue;

    const newArc = {
      centerX: center,
      centerY: center,
      start,
      end,
      filled,
      isComplete: filled === end,
      filledColor: segment.filledColor,
      emptyColor: segment.emptyColor,
      data: segment.data
    };

    arcs.push(newArc);
  });

  const lastFilledSegment = arcs.find(a => a.filled !== a.end) || arcs[arcs.length - 1];

  useEffect(() => {
    if (!lastFilledSegment) return;
    if (animationRunning.current) return;
    if (!isAnimated) return;
    animationRunning.current = true;
    Animated.timing(arcAnimatedValue, {
      toValue: lastFilledSegment.filled,
      duration: animationDuration,
      delay: animationDelay,
      useNativeDriver: false,
      easing: Easing.out(Easing.ease)
    }).start();

    const listenerId = arcAnimatedValue.addListener(e => {
      if (e.value === lastFilledSegment.filled) animationRunning.current = false;
    });
    return () => arcAnimatedValue.removeListener(listenerId);
  }, []);

  if (arcs.length === 0) {
    return null;
  }

  return (
    <View style={styles.container} testID="container">
      <Svg width={svgWidth} height={svgHeight}>
        <SegmentedArcContext.Provider
          value={{
            arcs,
            margin,
            center,
            filledArcWidth,
            radius,
            isAnimated,
            animationDuration,
            emptyArcWidth,
            totalArcs,
            arcsStart,
            spaceBetweenSegments,
            arcSegmentDegree,
            arcAnimatedValue,
            lastFilledSegment,
            ranges,
            rangesTextColor,
            rangesTextStyle,
            capInnerColor,
            capOuterColor,
            alignRangesWithSegments
          }}
        >
          {arcs.map((arc, index) => (
            <Segment key={index.toString()} arc={arc} />
          ))}

          <Cap />

          {showArcRanges && <RangesDisplay />}
        </SegmentedArcContext.Provider>
      </Svg>

      {children && <View style={localMiddleContentContainerStyle}>{children({ lastFilledSegment })}</View>}
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    alignItems: 'center'
  },
  rangeTextStyle: {
    fontSize: 12
  }
});

SegmentedArc.propTypes = {
  fillValue: PropTypes.number,
  segments: PropTypes.arrayOf(
    PropTypes.shape({
      scale: PropTypes.number,
      filledColor: PropTypes.string.isRequired,
      emptyColor: PropTypes.string.isRequired,
      data: PropTypes.object,
      arcDegreeScale: PropTypes.number
    })
  ).isRequired,
  filledArcWidth: PropTypes.number,
  emptyArcWidth: PropTypes.number,
  spaceBetweenSegments: PropTypes.number,
  arcDegree: PropTypes.number,
  radius: PropTypes.number,
  animationDuration: PropTypes.number,
  isAnimated: PropTypes.bool,
  animationDelay: PropTypes.number,
  showArcRanges: PropTypes.bool,
  children: PropTypes.func,
  middleContentContainerStyle: PropTypes.object,
  ranges: PropTypes.arrayOf(PropTypes.string),
  rangesTextColor: PropTypes.string,
  rangesTextStyle: PropTypes.object,
  capInnerColor: PropTypes.string,
  capOuterColor: PropTypes.string,
  alignRangesWithSegments: PropTypes.bool
};
export { SegmentedArcContext };
export default SegmentedArc;
