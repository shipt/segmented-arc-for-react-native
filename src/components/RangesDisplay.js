import React, { useContext, useMemo } from 'react';
import { G, Path, Defs, TextPath, Text } from 'react-native-svg';
import { SegmentedArcContext } from '../SegmentedArc';
import { drawArc } from '../utils/arcHelpers';

export const RangesDisplay = () => {
  const segmentedArcContext = useContext(SegmentedArcContext);
  const { radius, filledArcWidth, margin, arcsStart, arcSpacing, arcSize, ranges, rangesTextColor, rangesTextStyle } =
    segmentedArcContext;

  const { mappedPathKeys, rangesStartOffset } = useMemo(() => {
    return {
      mappedPathKeys: ranges.map((item, key) => `TextPath-${item}-${key}`),
      rangesStartOffset: ranges.map((_item, index) => `${index !== ranges.length - 1 ? 0 : 100}%`)
    };
  }, [ranges]);

  const _getRangesPath = (id, index) => {
    const rangeRadius = radius + filledArcWidth;
    const centerX = rangeRadius + margin;
    const centerY = rangeRadius + margin;
    const end = arcsStart + index * (arcSize + arcSpacing);
    const start = arcSize + end;

    return <Path key={id} id={id} d={drawArc(centerX, centerY, rangeRadius, start, end, true)} />;
  };

  const _getRangesDisplayValue = (key, index) => {
    let isLastRange = index === ranges.length - 1;
    let pathId = key;
    if (isLastRange) pathId = mappedPathKeys[index - 1];

    return (
      <Text key={key} fill={rangesTextColor} style={rangesTextStyle} textAnchor={isLastRange ? 'end' : 'start'}>
        <TextPath href={`#${pathId}`} startOffset={rangesStartOffset[index]}>
          {ranges[index]}
        </TextPath>
      </Text>
    );
  };
  return (
    <G>
      <Defs>{mappedPathKeys.map(_getRangesPath)}</Defs>
      {mappedPathKeys.map(_getRangesDisplayValue)}
    </G>
  );
};

export default RangesDisplay;
