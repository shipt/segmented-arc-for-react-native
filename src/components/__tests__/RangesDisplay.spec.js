import React from 'react';
import { RangesDisplay } from '../RangesDisplay';
import { SegmentedArcContext } from '../../SegmentedArc';
import { render } from '@testing-library/react-native';

describe('RangesDisplay', () => {
  let props = {};

  let wrapper;
  let contextValue = {};

  const getWrapper = properties => {
    return render(
      <SegmentedArcContext.Provider value={contextValue}>
        <RangesDisplay {...properties} />
      </SegmentedArcContext.Provider>
    );
  };

  beforeEach(() => {
    contextValue = {
      arcsStart: 0,
      arcSize: 180,
      radius: 100,
      filledArcWidth: 10,
      arcSpacing: 10,
      margin: 40,
      ranges: ['1.25', '2.5', '3.75', '5'],
      rangesTextColor: '#000'
    };
    props = {};

    wrapper = getWrapper(props);
  });

  it('renders default', () => {
    expect(wrapper).toMatchSnapshot();
  });
});
