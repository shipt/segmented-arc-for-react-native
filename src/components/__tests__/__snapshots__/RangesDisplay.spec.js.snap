// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RangesDisplay renders default 1`] = `
<RNSVGGroup>
  <RNSVGDefs>
    <RNSVGPath
      d="M 40 150 A 110 110 0 0 1 260 150"
      name="TextPath-1.25-0"
    />
    <RNSVGPath
      d="M 258.3288528313429 169.10129954336233 A 110 110 0 0 1 41.671147168657114 130.89870045663764"
      name="TextPath-2.5-1"
    />
    <RNSVGPath
      d="M 46.633811713550074 112.37778423417645 A 110 110 0 0 1 253.36618828644993 187.62221576582357"
      name="TextPath-3.75-2"
    />
    <RNSVGPath
      d="M 245.2627944162883 204.99999999999991 A 110 110 0 0 1 54.737205583711784 94.99999999999993"
      name="TextPath-5-3"
    />
  </RNSVGDefs>
  <RNSVGText
    content={null}
    dx={Array []}
    dy={Array []}
    fill={4278190080}
    font={
      Object {
        "textAnchor": "start",
      }
    }
    propList={
      Array [
        "fill",
      ]
    }
    rotate={Array []}
    x={Array []}
    y={Array []}
  >
    <RNSVGTextPath
      content={null}
      dx={Array []}
      dy={Array []}
      font={Object {}}
      href="TextPath-1.25-0"
      rotate={Array []}
      startOffset="0%"
      x={Array []}
      y={Array []}
    >
      <RNSVGTSpan
        content="1.25"
        dx={Array []}
        dy={Array []}
        font={Object {}}
        rotate={Array []}
        x={Array []}
        y={Array []}
      />
    </RNSVGTextPath>
  </RNSVGText>
  <RNSVGText
    content={null}
    dx={Array []}
    dy={Array []}
    fill={4278190080}
    font={
      Object {
        "textAnchor": "start",
      }
    }
    propList={
      Array [
        "fill",
      ]
    }
    rotate={Array []}
    x={Array []}
    y={Array []}
  >
    <RNSVGTextPath
      content={null}
      dx={Array []}
      dy={Array []}
      font={Object {}}
      href="TextPath-2.5-1"
      rotate={Array []}
      startOffset="0%"
      x={Array []}
      y={Array []}
    >
      <RNSVGTSpan
        content="2.5"
        dx={Array []}
        dy={Array []}
        font={Object {}}
        rotate={Array []}
        x={Array []}
        y={Array []}
      />
    </RNSVGTextPath>
  </RNSVGText>
  <RNSVGText
    content={null}
    dx={Array []}
    dy={Array []}
    fill={4278190080}
    font={
      Object {
        "textAnchor": "start",
      }
    }
    propList={
      Array [
        "fill",
      ]
    }
    rotate={Array []}
    x={Array []}
    y={Array []}
  >
    <RNSVGTextPath
      content={null}
      dx={Array []}
      dy={Array []}
      font={Object {}}
      href="TextPath-3.75-2"
      rotate={Array []}
      startOffset="0%"
      x={Array []}
      y={Array []}
    >
      <RNSVGTSpan
        content="3.75"
        dx={Array []}
        dy={Array []}
        font={Object {}}
        rotate={Array []}
        x={Array []}
        y={Array []}
      />
    </RNSVGTextPath>
  </RNSVGText>
  <RNSVGText
    content={null}
    dx={Array []}
    dy={Array []}
    fill={4278190080}
    font={
      Object {
        "textAnchor": "end",
      }
    }
    propList={
      Array [
        "fill",
      ]
    }
    rotate={Array []}
    x={Array []}
    y={Array []}
  >
    <RNSVGTextPath
      content={null}
      dx={Array []}
      dy={Array []}
      font={Object {}}
      href="TextPath-3.75-2"
      rotate={Array []}
      startOffset="100%"
      x={Array []}
      y={Array []}
    >
      <RNSVGTSpan
        content="5"
        dx={Array []}
        dy={Array []}
        font={Object {}}
        rotate={Array []}
        x={Array []}
        y={Array []}
      />
    </RNSVGTextPath>
  </RNSVGText>
</RNSVGGroup>
`;
